src/components/jscommon/CanvasWidget/CanvasWidgetNew.tsx

```diff
@@ -83,41 +83,46 @@ interface Props<T extends BaseLayerProps> {
+    const [prevDragState, setPrevDragState] = useState<DragState | null>(null)
+    const [prevWidth, setPrevWidth] = useState<number>(0)
+    const [prevHeight, setPrevHeight] = useState<number>(0)
 
     useEffect(() => {
-        // this is only needed if the previous repaint occurred before the canvas element was rendered to the browser
+        const divElement = divRef.current
         props.layers.forEach((L, index) => {
-            // /// TODO: figure out what to do here
-            // console.log(`I am layer ${index} and my props were ${JSON.stringify(props)}`)
-            const divElement = divRef.current
-            if (divElement) {
+            L.updateProps(props.widgetProps)
+            if ((divElement) && (divElement !== prevDivElement)) {
+                // only repaint if we have a new div element
                 const canvasElement = divElement.children[index]
                 L.resetCanvasElement(canvasElement)
-                L.updateProps(props.widgetProps)
-                L.scheduleRepaint() // The dependency list means this only gets called if props/canvas changed.
-                // In those cases, we definitely need to repaint, so scheduling it should be safe.
+                L.scheduleRepaint()
+            }
+            if ((props.widgetProps.width !== prevWidth) || (props.widgetProps.height !== prevHeight)) {
+                // repaint if dimensions have changed
+                L.scheduleRepaint()
             }
         })
-    }, [props, divRef])
-
-    const _dispatchDragEvents = useCallback((released: boolean) => {
-        if (!state.dragRect) return
-        const pixelDragRect = {
-            xmin: state.dragRect[0],
-            xmax: state.dragRect[0] + state.dragRect[2],
-            ymin: state.dragRect[1] + state.dragRect[3],
-            ymax: state.dragRect[1]
-        }
-        for (let l of props.layers) {
-            l.handleDrag(pixelDragRect, released, state.shift, state.dragAnchor, state.dragPosition)
+        if ((dragState) && (dragState !== prevDragState) && (dragState.dragRect)) {
+            // The drag state has changed, so we'll call the handleDrag on the layers
+            const pixelDragRect = {
+                xmin: dragState.dragRect[0],
+                xmax: dragState.dragRect[0] + dragState.dragRect[2],
+                ymin: dragState.dragRect[1] + dragState.dragRect[3],
+                ymax: dragState.dragRect[1]
+            }
+            for (let l of props.layers) {
+                l.handleDrag(pixelDragRect, !dragState.dragging, dragState.shift, dragState.dragAnchor, dragState.dragPosition)
+            }
         }
-    }, [props.layers, state.dragAnchor, state.dragRect, state.dragPosition, state.shift])
+        setPrevDivElement(divElement)
+        setPrevDragState(dragState)
+        setPrevWidth(props.widgetProps.width)
+        setPrevHeight(props.widgetProps.height)
```

Comment:
As you'd suggested, I was hoping to avoid having to do manual prior-state comparisons, but I guess they're still necessary... bummer. Is it because React is doing a shallow comparison on the `props` dependency, so it winds up 'changing' even if it doesn't actually change?